specs/lightclient/sync-protocol.md

```diff
@@ -173,11 +184,14 @@ def process_light_client_update(store: LightClientStore, update: LightClientUpda
     assert is_valid_light_client_update(store.snapshot, update)
     store.valid_updates.append(update)
 
-    if sum(update.sync_committee_bits) * 3 > len(update.sync_committee_bits) * 2 and update.header != update.finality_header:
+    if (
+        sum(update.sync_committee_bits) * 3 > len(update.sync_committee_bits) * 2
+        and update.header != update.finality_header
```

Comment:
what about the case where we reach the 2/3 quorum but do not have a finalized header?

it seems the intention here is to only immediately accept an update if we have quorum _and_ finality. in which case we want to make sure that `update.finality_header != BeaconBlockHeader()` instead of the second clause to the `and` as written (this conditional will be `True` if we have quorum and no finality as `update.header != BeaconBlockHeader()` and `update.finality_header == BeaconBlockHeader()`).

if we do not care about finality then we can remove the `and ...` and only check for quorum.

Reply:
Yeah, you are correct. Looks like a bug and should be `and update.finality_header != BeaconBlockHeader()` [which is a flag for no new finality]

Reply:
But I too question whether we *need* finality to update the light client store

Reply:
I think the light client being capable of reorging is reasonable. Definitely an okay response especially in cases of high network latency.

Reply:
This is making me realize that it would take some cleverness to make light clients capable of reorging. Here is one rough proposal. The LightClientStore keeps track of a _series_ of heads, which must always satisfy the following invariants:

* `store.heads[0]` must be finalized (if we want, if a block has been part of `store.heads` for >= 24 hours we can "finalize" it locally)
* `store.heads[n+1]` must be a descendant of `store.heads[n]`
* `vote_count(store.heads[n+1]) < vote_count(store.heads[n])`

Essentially, it is the client's best guess of the current chain, with confidence naturally decreasing as you get closer to the tip.

When a client receives a light client update, the update must reference an ancestor that is in `store.heads`. The update must satisfy the properties:

1. `votes(update.block) < votes(update.ancestor)`
2. `update.ancestor == store.heads[-1] or votes(update.block) > votes(store.heads[store.heads.index(update.ancestor) + 1])`

In other words, it must be a better alternative than whatever block it is replacing. If the update is accepted, then all elements from `store.heads` after `update.ancestor` are removed, and `update.block` is added to the end.

Reply:
Opened a new issue https://github.com/ethereum/eth2.0-specs/issues/2182 to discuss the light client reorg. 👍